// mighty-webcamjs/dist/components/SwitchCameraButton.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var React = require('react');
var PropTypes = require('prop-types');
var atom = require('atom-js');
var BaseComponent = require('./BaseComponent.js');

var SwitchCameraButton = function (_BaseComponent) {
	_inherits(SwitchCameraButton, _BaseComponent);

	function SwitchCameraButton() {
		var _ref;

		var _temp, _this, _ret;

		_classCallCheck(this, SwitchCameraButton);

		for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
			args[_key] = arguments[_key];
		}

		return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = SwitchCameraButton.__proto__ || Object.getPrototypeOf(SwitchCameraButton)).call.apply(_ref, [this].concat(args))), _this), _this.handleClick = function () {
			var webcam = _this.props.webcam;

			webcam.switchCamera().then(function () {
				webcam.reattach();
			});
		}, _temp), _possibleConstructorReturn(_this, _ret);
	}

	_createClass(SwitchCameraButton, [{
		key: 'render',
		value: function render() {
			return React.createElement(
				'button',
				{
					disabled: !this.props.live,
					className: this.props.cssPrefix + '__switch-camera-button',
					onClick: this.handleClick
				},
				this.props.switch_camera_node
			);
		}
	}]);

	return SwitchCameraButton;
}(BaseComponent);

SwitchCameraButton.propTypes = {
	cssPrefix: PropTypes.string.isRequired,
	live: PropTypes.bool.isRequired,
	model: PropTypes.object.isRequired,
	switch_camera_node: PropTypes.node.isRequired,
	webcam: PropTypes.object.isRequired
};


module.exports = atom.reactConnect('model', ['live', 'switch_camera_node'])(SwitchCameraButton);
